### ISSUE TITLE
channel stderr and stdout are not closed

### ISSUE THREAD

**Issue by MartinForReal** [go-cmd/cmd#26]
Please close stderr and stdout after cmd is executed
```go
for _,command := range task.Steps {
		// Create Cmd with options
		stepCmd := cmd.NewCmdOptions(cmdOptions,command.Cmd,command.Args...)
		task.Logger.Infof("Start executing %s %s ",command.Cmd, strings.Join(command.Args," "))
		flag := make(chan struct{},1)
		go func(flag chan struct{}) {
			for {
				select {
				case line,ok := <-stepCmd.Stdout:
					if ok {
						task.Logger.Infoln(line)
					} else {
						stepCmd.Stdout = nil
						task.Logger.Infof("Got all output from stdout")
					}
				case line,ok := <-stepCmd.Stderr:
					if ok {
						task.Logger.Errorln(line)
					} else {
						stepCmd.Stderr = nil
						task.Logger.Infof("Got all output from stderr")
					}
				}
				if stepCmd.Stderr == nil && stepCmd.Stdout == nil {
					task.Logger.Infof("Got all output")
					break
				}
			}
			flag <- struct{}{}
			close(flag)
		}(flag)
		// Run and wait for Cmd to return, discard Status
		status:= <-stepCmd.Start()
		task.Logger.Infof("Command '%s %s' is executed with return code %d",command.Cmd, strings.Join(command.Args," ") , status.Exit)
		<-flag
```
the code above will never return

**Comment by daniel-nichter** [comment]
The streaming chans are not closed by design. Problem is: we cannot know if/when they're safe to close, where "safe" means "caller has read all lines". The chans are a buffer (of `DEFAULT_STREAM_CHAN_SIZE` size). So this sequence of events can happen:

1. Caller starts recv on chan
1. Caller starts cmd
1. Cmd prints "a\n"
1. Cmd prints "b\n"
1. Cmd prints "c\n"
1. Cmd exits
1. Caller reads first line ("a\n")

And that's it. If we close the chan after 6, the caller loses some or all of the output.

This is documented:
> Since the channel is not closed by the OutputStream, the two indications that all lines have been sent and received are the command finishing and the channel size being zero.

Let me know if I'm missing something, or if there's a safe way to close the chan.

**Comment by malexdev** [comment]
>If we close the chan after 6, the caller loses some or all of the output.

Channels don't work like that: a `close` on a channel is just another message, so it gets buffered in line like everything else.
For demonstration, refer to this example: https://goplay.space/#B6Kw81F-T1Z

This is also documented in the [Go reference](https://golang.org/ref/spec#Close) (emphasis mine):
```
After calling close, and *after any previously sent values have been received*, receive operations will return the zero value for the channel's type without blocking.
```

**Comment by daniel-nichter** [comment]
Re-reading my last comment, I see I didn't explain it well. My concern was not with the channel but with the program sending to the channel. My original understanding was that Go os/exec does not wait for stdout/stderr. But re-reading [cmd.Wait](https://golang.org/pkg/os/exec/#Cmd.Wait) today it's clear:

> Wait waits for the command to exit and waits for any copying to stdin or copying from stdout or stderr to complete. ... If any of c.Stdin, c.Stdout or c.Stderr are not an *os.File, Wait also waits for the respective I/O loop copying to or from the process to complete.

These docs changed at some point. Today, the [exec.Cmd](https://golang.org/pkg/os/exec/#Cmd) docs for `Stdout io.Writer` and `Stderr io.Writer` say,

> In this case, Wait does not complete until the goroutine reaches EOF or encounters an error.

but previous versions did not say that. Another example from Cmd.Wait docs:

> If any of c.Stdin, c.Stdout or c.Stderr are not an *os.File, Wait also waits for the respective I/O loop copying to or from the process to complete.

vs.

> If c.Stdin is not an *os.File, Wait also waits for the I/O loop copying from c.Stdin into the process's standard input to complete.

Former is today, latter is go1.9. No mention of Stdout and Stderr. I wish the docs said if changes are purely doc-related or also code-related because this makes me wonder: did Wait start to wait for stdout/stderr output as of some version?

With go1.11 it's clear: yes, we can close the chan without any data loss. But for go1.9 we'll need to test it.

**Comment by rasharab** [comment]
As of January 2020, can we introduce closing these channels?
I'd like to move to using waitGroup semantics rather than doing len queries.

**Comment by daniel-nichter** [comment]
Agreed. Will implement and release as v1.2.0. Even though current v1.1.0 doesn't have any additional releases, this change is probably important enough to warrant a new minor level, and to make sure no one using v1.1 is surprised by the channel being closed.

**Comment by daniel-nichter** [comment]
Released as v1.2.0.
